### ISSUE TITLE
Ignores empty rules without warning

### ISSUE THREAD

**Issue by antifuchs** [thelegy/nixos-nftables-firewall#10]
So today I brought my new firewall online using this module, and it was really nice and easy to configure! Unfortunately it seems to have a severe bug, where a forwarding rule doesn't get emitted - which unfortunately is the one that every zone's internet egress work.

Here's a configuration that triggers that problem:

```nix
    networking.nftables.firewall = {
      enable = true;
      zones = {
        # WAN interface:
        theWorld.interfaces = ["enp0s20f0"];

        egressable.interfaces = [
          "clients"
        ];

        something.interfaces = ["toys"];
      };

      rules = {
        communication = {
          from = ["egressable"];
          to = ["something"];
        };

        egress = {
          from = ["egressable"];
          to = ["theWorld"];
        };
      };
    };
```

If you generate this configuration, you'll notice that no chain or inlined rule gets generated that corresponds in any way to `egress`. 

Here's the output I get when retrieving `nixosConfigurations.leigh.config.networking.nftables.ruleset` from my flake:

```nix
nix-repl> nixosConfigurations.leigh.config.networking.nftables.ruleset
"table inet firewall {\n\n  chain forward {\n    type filter hook forward priority 0; policy drop;\n    ct state {established, related} accept\n    ct state invalid drop\n    counter drop\n  }\n\n  chain input {\n    type filter hook input priority 0; policy drop\n    iifname { lo } accept\n    ct state {established, related} accept\n    ct state invalid drop\n    jump traverse-from-all-zone-to-fw-zone-rule\n    counter drop\n  }\n\n  chain postrouting {\n    type nat hook postrouting priority srcnat;\n  }\n\n  chain prerouting {\n    type nat hook prerouting priority dstnat;\n  }\n\n  chain rule-icmp {\n    ip6 nexthdr icmpv6 icmpv6 type { echo-request, nd-router-advert, nd-neighbor-solicit, nd-neighbor-advert } accept\n    ip protocol icmp icmp type { echo-request, router-advertisement } accept\n    ip6 saddr fe80::/10 ip6 daddr fe80::/10 udp dport 546 accept\n  }\n\n  chain rule-nixos-firewall {\n    tcp dport { 22 } accept\n    udp dport { 5353 } accept\n  }\n\n  chain traverse-from-all-zone-to-fw-zone-rule {\n    tcp dport { 22 } accept  # inlined: rule-ssh\n    jump rule-icmp\n    jump rule-nixos-firewall\n  }\n\n}\n"
```

I tried some variations of the above config, too:
* Other rules touching other zones seem to correctly get emitted, including ones that look like "communication".
* Other rules touching *one of these* zones seem to correctly get emitted.
* This rule or its inverse (from=theWorld, to=egressable; nonsense, but I tried it) get dropped.
* Another time, I tried an equivalent-looking configuration above and both the "communication" and "egress" rules were emitted. I'm completely doubting my sanity right now.

Anyway - for now, I solved this by manually generating a chain entry with the corresponding nftables rule, but this is not ideal & smells pretty dangerous to me.

**Comment by antifuchs** [comment]
Oh god, I think I understand now what's wrong: There is no explicit verdict set on those rules, which means it defaults to `null`, which has somewhat murky [definition](https://thelegy.github.io/nixos-nftables-firewall/modules/nftables-zoned/#networking-nftables-firewall-zones-name-parent):

> If traffic matches a zone, it will first be tested, if it also matches some of its subzones. If so, the logic of the subzones will be called. If not, or if the subzones did not terminate the rule processing with a verdict, the rules of the parent zone will be applied.

I'm not sure if I read this right, but it sounds like the expectation is that the rule with a `null` verdict will apply to a zone whose parent has a rule with a real verdict (or at least extraLines); if not, that chain will be optimized away.

I would love if the module could emit a warning about an ineffective rule there.

**Comment by thelegy** [comment]
Hi there, sorry for the wait. Your quote has little to do with what you are observing here. It describes, what happens, if you nest zones. None of the zones in your example has a `parent`, so no zone nesting is happening here.

But you are right, that you are not specifying what your rule should do here. I am biased, because I wrote it, so maybe you can tell me what you had expected it to generate? Drop or accept traffic?

In any case, maybe I can actually detect if a rule has no verdict, no allowed tcp/udp ports, no port ranges and no `extraLines` and indeed emit a warning. Defaulting to `accept` seems pretty unsafe to me and defaulting to `drop` also does not seem right to me.

**Comment by antifuchs** [comment]
> But you are right, that you are not specifying what your rule should do here. I am biased, because I wrote it, so maybe you can tell me what you had expected it to generate? Drop or accept traffic?

I was expecting it to accept traffic, as the policy is to drop; I'd expected that I go to the trouble of listing a set of zones would indicate to the system that allowing them would be desired. Now that I've used the system more, the behavior makes sense. However, on occasion I still get caught in a state where I think I listed the things I need but the system doesn't work - it always turns out that I'd forgotten an `extraLines` or port list in that case. My feeling is that forgetting a directive should fail the compile.
